sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala

```diff
@@ -394,4 +403,19 @@ class DatasetAggregatorSuite extends QueryTest with SharedSparkSession {
     checkAnswer(group, Row("bob", Row(true, 3)) :: Nil)
     checkDataset(group.as[OptionBooleanIntData], OptionBooleanIntData("bob", Some((true, 3))))
   }
+
+  test("SPARK-30590: untyped select should not accept typed column without input type") {
+    val df = Seq((1, 2, 3, 4, 5, 6)).toDF("a", "b", "c", "d", "e", "f")
+    val fooAgg = (i: Int) => FooAgg(i).toColumn.name(s"foo_agg_$i")
+
+    val agg1 = df.select(fooAgg(1), fooAgg(2), fooAgg(3), fooAgg(4), fooAgg(5))
+    checkDataset(agg1, (3, 5, 7, 9, 11))
+
+    // Passes typed columns to untyped `Dataset.select` API.
+    val err = intercept[AnalysisException] {
+      df.select(fooAgg(1), fooAgg(2), fooAgg(3), fooAgg(4), fooAgg(5), fooAgg(6))
```

Comment:
Not related to this PR, just a note:

We have 5 overloads of typed `select`, and typed `count` is supported in both typed and untyped `select`. That said, if we add a 6th overload of typed `select`, it can break queries that call the untyped `select` with 6 typed `count`s.

I'm not sure what's the best way to move forward. Maybe we should add new methods `typedSelect` to disambiguate the untyped version.

Reply:
Yea, to be clear, if we add a 6th overload of typed `select`, a call to the untyped `select` with 6 typed `count` could return `Dataset[(Long, Long, ...)]` instead of `DataFrame`.

I think you meant something like existing `selectUntyped`? Although its naming is confusing.